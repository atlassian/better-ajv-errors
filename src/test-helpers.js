import { readFileSync } from 'fs';
import { getFixturePath } from 'jest-fixtures';
import Ajv from 'ajv';

export async function getSchemaAndData(name, dirPath) {
  const schemaPath = await getFixturePath(dirPath, name, 'schema.json');
  const schema = JSON.parse(readFileSync(schemaPath, 'utf8'));
  const dataPath = await getFixturePath(dirPath, name, 'data.json');
  const data = JSON.parse(readFileSync(dataPath, 'utf8'));

  return [schema, data];
}

export function evaluateSchema(schema, data) {
  const ajv = new Ajv();
  const fn = ajv.compile(schema);
  fn(data);
  expect(fn.errors.length).toBeGreaterThanOrEqual(0);
  return fn.errors;
}
