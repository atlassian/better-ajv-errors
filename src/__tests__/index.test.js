import Ajv from 'ajv';
import Ajv2020 from 'ajv/dist/2020';
import { getSchemaAndData } from '../test-helpers';
import betterAjvErrors from '..';
import betterAjvErrorsBabelExport from '../../lib';
import { openapi } from '@apidevtools/openapi-schemas';

describe('Main', () => {
  it.each([
    ['should output error with reconstructed codeframe', true],
    ['should output error with reconstructed codeframe [without colors]', false],
  ])('%s', async (_, colorize) => {
    const [schema, data] = await getSchemaAndData('default', __dirname);
    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      colorize,
      format: 'cli',
      indent: 2,
    });
    expect(res).toMatchSnapshot();
  });

  it('should output error with codeframe', async () => {
    const [schema, data, json] = await getSchemaAndData('default', __dirname);
    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'cli',
      json,
    });
    expect(res).toMatchSnapshot();
  });

  describe('Babel export', () => {
    it('should function as normal', async () => {
      const [schema, data] = await getSchemaAndData('default', __dirname);
      const ajv = new Ajv();
      const validate = ajv.compile(schema);
      const valid = validate(data);
      expect(valid).toBe(false);

      const res = betterAjvErrorsBabelExport(schema, data, validate.errors);
      expect(res).toMatchSnapshot();
    });
  });

  describe('complex schema examples', () => {
    it('should output an error on an invalid OpenAPI 3.1 definition', async () => {
      const schema = openapi.v31;
      const [, data] = await getSchemaAndData('openapi-3.1', __dirname);

      // Need to load the 2020 dist because it supports draft-7, which we'll need for an 3.1.0
      // OpenAPI definition.
      const ajv = new Ajv2020({ allErrors: true, strict: false, validateFormats: false });

      const validate = ajv.compile(schema);
      const valid = validate(data);
      expect(valid).toBe(false);

      const res = betterAjvErrors(schema, data, validate.errors, {
        indent: 2,
        format: 'cli',
      });
      expect(res).toMatchSnapshot();
    });
  });
});
