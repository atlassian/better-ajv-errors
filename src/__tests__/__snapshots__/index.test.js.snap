// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Babel export should function as normal 1`] = `
"[31m[1mENUM[22m[39m[31m must be equal to one of the allowed values[39m
[31m(paragraph, codeBlock, blockquote)[39m

[0m[31m[1m>[22m[39m[90m 1 |[39m {[32m\\"type\\"[39m[33m:[39m[32m\\"doc\\"[39m[33m,[39m[32m\\"version\\"[39m[33m:[39m[35m1[39m[33m,[39m[32m\\"content\\"[39m[33m:[39m[{[32m\\"type\\"[39m[33m:[39m[32m\\"paragarph\\"[39m}]}[0m
[0m [90m   |[39m                                              [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mparagraph[31m here?[22m[39m[0m"
`;

exports[`Main complex schema examples should output an error on an invalid OpenAPI 3.1 definition 1`] = `
"[31m[1mUNEVALUATEDPROPERTIES[22m[39m[31m must NOT have unevaluated properties[39m

[0m [90m 12 |[39m   [32m\\"paths\\"[39m[33m:[39m {[0m
[0m [90m 13 |[39m     [32m\\"/anything\\"[39m[33m:[39m {[0m
[0m[31m[1m>[22m[39m[90m 14 |[39m       [32m\\"get\\"[39m[33m:[39m {[0m
[0m [90m    |[39m              [31m[1m^[22m[39m [31m[1m👈🏽  [95munevaluatedProperties[31m must NOT have unevaluated properties[22m[39m[0m
[0m [90m 15 |[39m         [32m\\"summary\\"[39m[33m:[39m [32m\\"Example operation\\"[39m[33m,[39m[0m
[0m [90m 16 |[39m         [32m\\"description\\"[39m[33m:[39m [32m\\"This operation has \`tags\` typod.\\"[39m[33m,[39m[0m
[0m [90m 17 |[39m         [32m\\"tagss\\"[39m[33m:[39m [[0m"
`;

exports[`Main should output error with codeframe 1`] = `
"[31m[1mENUM[22m[39m[31m must be equal to one of the allowed values[39m
[31m(paragraph, codeBlock, blockquote)[39m

[0m [90m 2 |[39m   [32m\\"type\\"[39m[33m:[39m [32m\\"doc\\"[39m[33m,[39m[0m
[0m [90m 3 |[39m   [32m\\"version\\"[39m[33m:[39m [35m1[39m[33m,[39m[0m
[0m[31m[1m>[22m[39m[90m 4 |[39m   [32m\\"content\\"[39m[33m:[39m [{ [32m\\"type\\"[39m[33m:[39m [32m\\"paragarph\\"[39m }][0m
[0m [90m   |[39m                         [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mparagraph[31m here?[22m[39m[0m
[0m [90m 5 |[39m }[0m
[0m [90m 6 |[39m[0m"
`;

exports[`Main should output error with reconstructed codeframe [without colors] 1`] = `
"ENUM must be equal to one of the allowed values
(paragraph, codeBlock, blockquote)

  4 |   \\"content\\": [
  5 |     {
> 6 |       \\"type\\": \\"paragarph\\"
    |               ^^^^^^^^^^^ 👈🏽  Did you mean paragraph here?
  7 |     }
  8 |   ]
  9 | }"
`;

exports[`Main should output error with reconstructed codeframe 1`] = `
"[31m[1mENUM[22m[39m[31m must be equal to one of the allowed values[39m
[31m(paragraph, codeBlock, blockquote)[39m

[0m [90m 4 |[39m   [32m\\"content\\"[39m[33m:[39m [[0m
[0m [90m 5 |[39m     {[0m
[0m[31m[1m>[22m[39m[90m 6 |[39m       [32m\\"type\\"[39m[33m:[39m [32m\\"paragarph\\"[39m[0m
[0m [90m   |[39m               [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mparagraph[31m here?[22m[39m[0m
[0m [90m 7 |[39m     }[0m
[0m [90m 8 |[39m   ][0m
[0m [90m 9 |[39m }[0m"
`;
